README.rst
============================
reStructuredText with Sphinx
============================
.. class:: no-web no-pdf

    |npm| |build| |test| |coverage| |vulnerabilities|

.. |npm| image:: https://img.shields.io/npm/v/@passoa/cxb
    :target: https://www.npmjs.com/package/@passoa/cxb
    :alt: npm
.. |build| image:: https://img.shields.io/azure-devops/build/passoa/cxb/33
    :target: https://dev.azure.com/passoa/cxb/_build/latest?definitionId=33
    :alt: Azure DevOps tests
.. |test| image:: https://img.shields.io/azure-devops/tests/passoa/cxb/33
    :target: https://dev.azure.com/passoa/cxb/_build/latest?definitionId=33
    :alt: Azure DevOps tests
.. |coverage| image:: https://img.shields.io/azure-devops/coverage/passoa/cxb/33   
    :target: https://dev.azure.com/passoa/cxb/_build/latest?definitionId=33
    :alt: Azure DevOps tests
.. |vulnerabilities| image:: https://img.shields.io/snyk/vulnerabilities/npm/@passoa/cxb
    :alt: Snyk Vulnerabilities for npm scoped package

-----------------
cxb
-----------------

个人自用的一个node扩展编译工具，它参考了 `cmake-js <https://github.com/cmake-js/cmake-js>`_ 的代码，目标是可以适配node和私有js引擎平台（如arm平台下的duktape，quickjs等）。cxb可以配合napi或者node-api-addon编译node扩展，也可以配合私有sdk编译arm下quickjs或duktape的cpp扩展。

-----------------
安装
-----------------


* 使用npm

     npm install @passoa/cxb

-----------------
帮助
-----------------

    cxb --help

-----------------
依赖要求
-----------------

* 安装 `cmake <http://www.cmake.org/download/>`_
* 安装 C/C++ 编译工具链

  - gcc
  - clang
  - msvc
  - mingw
  - 其他被cmake支持的编译工具链

-----------------
使用方法
-----------------

暂无，可以参考test/simple，它提供了一个简单的node-api-addon编译例子。

---------------------
文档（暂无）
---------------------

* `documentation <http://cxb.readTheDocs.org/>`_

-----
使用例子（暂无）
-----

* `example <https://github.com/pass0a/cxbtest>`_
https://www.youtube.com/watch?v=qexy4Ph66JE

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.14](https://github.com/pass0a/cvip/compare/v1.0.13...v1.0.14) (2020-05-15)


### Bug Fixes

* 修复类型文件不匹配的问题，去掉摄像头命名，增加isOpened接口,并修改azurepipelines的编译流程 ([cccaf15](https://github.com/pass0a/cvip/commit/cccaf15a0b82a7b293f90dfcbcd5e344dcadc985))

### [1.0.13](https://github.com/pass0a/cvip/compare/v1.0.12...v1.0.13) (2020-05-14)


### Bug Fixes

* 修改保存摄像头数据为图片文件，修改摄像头原始内容转换为jpg内容 ([fa42533](https://github.com/pass0a/cvip/commit/fa4253313987f053a277a362f3bd4a7ae907505b))

### [1.0.12](https://github.com/pass0a/cvip/compare/v1.0.11...v1.0.12) (2020-05-08)


### Bug Fixes

* 增加npm cache功能，更新cxb版本 ([01125d8](https://github.com/pass0a/cvip/commit/01125d8a32035ac17c652048b55ee58f80805de7))
* 增加关闭摄像头的接口，并修复调用接口不一致的问题 ([4d4f288](https://github.com/pass0a/cvip/commit/4d4f28879936db29735fba148e98ed0fde9fa52d))

### [1.0.12](https://github.com/pass0a/cvip/compare/v1.0.11...v1.0.12) (2020-05-08)


### Bug Fixes

* 增加npm cache功能，更新cxb版本 ([0a6c464](https://github.com/pass0a/cvip/commit/0a6c464448abfc56a178febd52ac8e81e629f8f4))
* 增加关闭摄像头的接口，并修复调用接口不一致的问题 ([4d4f288](https://github.com/pass0a/cvip/commit/4d4f28879936db29735fba148e98ed0fde9fa52d))

### [1.0.11](https://github.com/pass0a/cvip/compare/v1.0.10...v1.0.11) (2020-05-08)


### Features

* 增加多路摄像头的功能 ([8aa3660](https://github.com/pass0a/cvip/commit/8aa3660cc373e21eb13822906e21cbdcb3ac9bbf))


### Bug Fixes

* 修复压缩包文件错误的问题，修复opencv依赖dll没有拷贝的问题 ([0ada84c](https://github.com/pass0a/cvip/commit/0ada84c50db87a50e0ff18ad45de481fc322dd7e))
* 修复在azurepipelines上无法编译通过的步骤，并增加单元测试报告 ([6a4a05e](https://github.com/pass0a/cvip/commit/6a4a05e7aa332d6151839f4e7225ded53bbf432e))
* 增加camera功能，增加jest测试框架，更新cxb设置 ([6be8365](https://github.com/pass0a/cvip/commit/6be8365ce9366e97f76c7f103668c68b65578c3d))

### [1.0.10](https://github.com/pass0a/cvip/compare/v1.0.9...v1.0.10) (2020-03-20)


### Bug Fixes

* 修改远程下载的包版本包没有包含引导字符v的问题 ([fbe2ab5](https://github.com/pass0a/cvip/commit/fbe2ab519011d429bcb75ec3859dffc7a985c2f8))

### [1.0.9](https://github.com/pass0a/cvip/compare/v1.0.7...v1.0.9) (2020-03-19)


### Bug Fixes

* 不指定平台编译，方便根据当前平台选择不同的二进制包 ([3f370b4](https://github.com/pass0a/cvip/commit/3f370b4bfa73afcd93ba7ca603a2681506264b8a))

### [1.0.10](https://github.com/pass0a/cvip/compare/v1.0.7...v1.0.10) (2020-03-19)


### Bug Fixes

*  不指定平台编译，方便根据当前平台选择不同的二进制包 ([59a2903](https://github.com/pass0a/cvip/commit/59a2903c8d45c318f42cc0eb4217592b9a4660e9))

### [1.0.9](https://github.com/pass0a/cvip/compare/v1.0.7...v1.0.9) (2020-03-19)


### Bug Fixes

* 不指定平台编译，方便根据当前平台选择不同的二进制包 ([a9b3bb9](https://github.com/pass0a/cvip/commit/a9b3bb96aa5896864a34b969dd41e1cf1e3d74f8))

### [1.0.7](https://github.com/pass0a/cvip/compare/v1.0.6...v1.0.7) (2020-03-18)


### Features

* upgrade opencv 3 to 4 ([85777bb](https://github.com/pass0a/cvip/commit/85777bb947c223e89e405dc4b636fb32e9a71948))


### Bug Fixes

* 解决拷贝opencv的dll版本不一致的问题 ([1bbf829](https://github.com/pass0a/cvip/commit/1bbf8296fbdecf4aec83cd924c7807eb41988cd2))

### [1.0.6](https://github.com/pass0a/cvip/compare/v1.0.4...v1.0.6) (2020-03-16)


### Features

* 合并回主分支，增加node版本 ([59fa190](https://github.com/pass0a/cvip/commit/59fa190eca673459432e6ccc6979fa9ec5b3233c))
* 实现匹配子图，重新编码和截图功能 ([e664c35](https://github.com/pass0a/cvip/commit/e664c353b8d4b8d550e0dfa79fb15aa0281b8e3c))


src/image.ts
import { bindings } from './bindings';

export let jpeg = 1;
export let webp = 64;
export let png = 16;
export function imageMatch(img: string, tmp: string): { valid: number; x: number; y: number; val: number } {
	return bindings.cvipMatch(img, tmp);
}
export function imageSave(tmp: string, sim: string, cvfmt: number): number {
	return bindings.cvipEncode(tmp, sim, cvfmt);
}
export function imageCut(tmp: string, sim: string, cvfmt: number, x: number, y: number, w: number, h: number): number {
	return bindings.cvipCut(tmp, sim, cvfmt, x, y, w, h);
}
export function stdev() {
	//return cvip.stdev.apply(this, arguments);
}
export function absdiff() {
	//return cvip.absdiff.apply(this, arguments);
}
export function horizontal() {
	//return cvip.horizontal.apply(this, arguments);
}
export function resave() {
	//return cvip.resave.apply(this, arguments);
}
export function cutcap() {
	//return cvip.cutcap.apply(this, arguments);
}

export enum CV_PROP {
	CAP_PROP_POS_MSEC = 0, //!< Current position of the video file in milliseconds.
	CAP_PROP_POS_FRAMES = 1, //!< 0-based index of the frame to be decoded/captured next.
	CAP_PROP_POS_AVI_RATIO = 2, //!< Relative position of the video file: 0=start of the film, 1=end of the film.
	CAP_PROP_FRAME_WIDTH = 3, //!< Width of the frames in the video stream.
	CAP_PROP_FRAME_HEIGHT = 4, //!< Height of the frames in the video stream.
	CAP_PROP_FPS = 5, //!< Frame rate.
	CAP_PROP_FOURCC = 6, //!< 4-character code of codec. see VideoWriter::fourcc .
	CAP_PROP_FRAME_COUNT = 7, //!< Number of frames in the video file.
	CAP_PROP_FORMAT = 8, //!< Format of the %Mat objects returned by VideoCapture::retrieve().
	CAP_PROP_MODE = 9, //!< Backend-specific value indicating the current capture mode.
	CAP_PROP_BRIGHTNESS = 10, //!< Brightness of the image (only for those cameras that support).
	CAP_PROP_CONTRAST = 11, //!< Contrast of the image (only for cameras).
	CAP_PROP_SATURATION = 12, //!< Saturation of the image (only for cameras).
	CAP_PROP_HUE = 13, //!< Hue of the image (only for cameras).
	CAP_PROP_GAIN = 14, //!< Gain of the image (only for those cameras that support).
	CAP_PROP_EXPOSURE = 15, //!< Exposure (only for those cameras that support).
	CAP_PROP_CONVERT_RGB = 16, //!< Boolean flags indicating whether images should be converted to RGB.
	CAP_PROP_WHITE_BALANCE_BLUE_U = 17, //!< Currently unsupported.
	CAP_PROP_RECTIFICATION = 18, //!< Rectification flag for stereo cameras (note: only supported by DC1394 v 2.x backend currently).
	CAP_PROP_MONOCHROME = 19,
	CAP_PROP_SHARPNESS = 20,
	CAP_PROP_AUTO_EXPOSURE = 21, //!< DC1394: exposure control done by camera, user can adjust reference level using this feature.
	CAP_PROP_GAMMA = 22,
	CAP_PROP_TEMPERATURE = 23,
	CAP_PROP_TRIGGER = 24,
	CAP_PROP_TRIGGER_DELAY = 25,
	CAP_PROP_WHITE_BALANCE_RED_V = 26,
	CAP_PROP_ZOOM = 27,
	CAP_PROP_FOCUS = 28,
	CAP_PROP_GUID = 29,
	CAP_PROP_ISO_SPEED = 30,
	CAP_PROP_BACKLIGHT = 32,
	CAP_PROP_PAN = 33,
	CAP_PROP_TILT = 34,
	CAP_PROP_ROLL = 35,
	CAP_PROP_IRIS = 36,
	CAP_PROP_SETTINGS = 37, //!< Pop up video/camera filter dialog (note: only supported by DSHOW backend currently. The property value is ignored)
	CAP_PROP_BUFFERSIZE = 38,
	CAP_PROP_AUTOFOCUS = 39,
	CAP_PROP_SAR_NUM = 40, //!< Sample aspect ratio: num/den (num)
	CAP_PROP_SAR_DEN = 41, //!< Sample aspect ratio: num/den (den)
	CAP_PROP_BACKEND = 42, //!< Current backend (enum VideoCaptureAPIs). Read-only property
	CAP_PROP_CHANNEL = 43, //!< Video input or Channel Number (only for those cameras that support)
	CAP_PROP_AUTO_WB = 44, //!< enable/ disable auto white-balance
	CAP_PROP_WB_TEMPERATURE = 45 //!< white-balance color temperature
}
